src/core/types/type_categorical.cc

```diff
@@ -0,0 +1,145 @@
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "stype.h"
+#include "types/type_invalid.h"
+#include "types/type_categorical.h"
+#include "utils/assert.h"
+namespace dt {
+
+
+//------------------------------------------------------------------------------
+// Type_Cat
+//------------------------------------------------------------------------------
+
+Type_Cat::Type_Cat(SType st, Type t) : TypeImpl(st) {
+  if (t.is_categorical()) {
+    throw TypeError()
+      << "Categories are not allowed to be of a categorical type";
+  }
+  elementType_ = std::move(t);
+}
+
+bool Type_Cat::is_compound() const {
+  return true;
+}
+
+bool Type_Cat::is_categorical() const {
+  return true;
+}
+
+bool Type_Cat::can_be_read_as_int8() const {
+  return elementType_.can_be_read_as<int8_t>();
+}
+
+bool Type_Cat::can_be_read_as_int16() const {
+  return elementType_.can_be_read_as<int16_t>();
+}
+
+bool Type_Cat::can_be_read_as_int32() const {
+  return elementType_.can_be_read_as<int32_t>();
+}
+
+bool Type_Cat::can_be_read_as_int64() const {
+  return elementType_.can_be_read_as<int64_t>();
+}
+
+bool Type_Cat::can_be_read_as_float32() const {
+  return elementType_.can_be_read_as<float>();
+}
+
+bool Type_Cat::can_be_read_as_float64() const {
+  return elementType_.can_be_read_as<double>();
+}
+
+bool Type_Cat::can_be_read_as_cstring() const {
+  return elementType_.can_be_read_as<CString>();
+}
+
+bool Type_Cat::can_be_read_as_pyobject() const {
+  return elementType_.can_be_read_as<py::oobj>();
+}
+
+bool Type_Cat::can_be_read_as_column() const {
+  return elementType_.can_be_read_as<Column>();
+}
+
+
+TypeImpl* Type_Cat::common_type(TypeImpl* other) {
+  if (other->is_categorical() &&
+      elementType_ == reinterpret_cast<const Type_Cat*>(other)->elementType_)
+  {
+    return other->stype() > stype() ? other : this;
+  }
+  if (other->is_void()) {
+    return this;
+  }
+  if (other->is_object() || other->is_invalid()) {
+    return other;
+  }
+  return new Type_Invalid();
+}
```

Comment:
Yes, there are many cases that we should take into account here. For the moment I made this as simple as possible and will expand this method in the next PRs (by also updating the others — like to `cat(T), T` we need to update the `common_type` method for `T`).

Reply:
you could always try something like `if (other.is_compound()) return other->common_type(this);`

Reply:
but yeah, if you feel it would significantly complicate the PR then let's leave it as-is for now

Reply:
>you could always try something like if (other.is_compound()) return other->common_type(this);

Yes, can do something like this. But still need to go through all of the existing types for that. Better to have a separate PR since have to touch many files that are not related to categoricals.